src/Features/Core/Portable/Completion/CommitRule.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    /// <summary>
+    /// A <see cref="CommitRule"/> describes how a completion controller 
+    /// </summary>
+    internal class CommitRule
+    {
+        /// <summary>
+        /// The kind of commit rules.
+        /// </summary>
+        public CommitRuleKind Kind { get; }
+
+        /// <summary>
+        /// The key characters associated with the rule.
+        /// This is one or more characters associated with key presses.
+        /// </summary>
+        public string Keys { get; }
+
+        /// <summary>
+        /// Optional match string; usage depends on the rule.
+        /// </summary>
+        public string Match { get; }
+
+        /// <summary>
+        /// True if matches are case sensitive.
+        /// </summary>
+        public bool IsCaseSensitive { get; }
+
+        private CommitRule(CommitRuleKind kind, string keys, string match, bool isCaseSensitive)
+        {
+            this.Kind = kind;
+            this.Keys = keys;
+            this.Match = match;
+            this.IsCaseSensitive = isCaseSensitive;
+        }
+
+        public static CommitRule Create(CommitRuleKind kind, string keys)
+        {
+            return new CommitRule(kind, keys, null, isCaseSensitive: false);
+        }
+
+        public static CommitRule Create(CommitRuleKind kind, string keys, string match, bool isCaseSensitive)
+        {
+            return new CommitRule(kind, keys, match, isCaseSensitive);
+        }
+    }
+
+    internal enum CommitRuleKind
+    {
+        /// <summary>
+        /// The <see cref="CommitRule"/> specifies the set of keys that are used to 
+        /// commit the <see cref="CompletionItem"/> instead of the default keys.
+        /// </summary>
+        ReplaceKeys,
+
+        /// <summary>
+        /// The <see cref="CommitRule"/> specifies additional keys that can be used to
+        /// commit the <see cref="CompletionItem"/>.
+        /// </summary>
+        IncludeKeys,
+
+        /// <summary>
+        /// The <see cref="CommitRule"/> specifies keys that cannot be used to 
+        /// commit the <see cref="CompletionItem"/>
+        /// </summary>
+        ExcludeKeys,
+
+        /// <summary>
+        /// The <see cref="CommitRule"/> specifies keys that cannot be used to 
+        /// commit the <see cref="CompletionItem"/> if the text typed since the start of completion
+        /// ends with the <see cref="CommitRule.Match"/> text.
+        /// </summary>
+        ExcludeKeysIfMatchEndOfTypedText
```

Comment:
Skeptical if this is necessary.  Even for the ```\\``` case in interactive  I'd prefer this just be gone.